core/drivers/firewall/stm32_risaf.c

```diff
@@ -0,0 +1,770 @@
+	     io_read32(base + _RISAF_REG_CIDCFGR(region_id)));
+
+	return TEE_SUCCESS;
+}
+
+static void risaf_print_version(struct stm32_risaf_instance *risaf)
+{
+	vaddr_t base = risaf_base(risaf);
+	struct stm32_risaf_version __maybe_unused version = {
+		.major = (io_read32(base + _RISAF_VERR) &
+			  _RISAF_VERR_MAJREV_MASK) >> _RISAF_VERR_MAJREV_SHIFT,
+		.minor = (io_read32(base + _RISAF_VERR) &
+			  _RISAF_VERR_MINREV_MASK) >> _RISAF_VERR_MINREV_SHIFT,
+		.ip_id = io_read32(base + _RISAF_IPIDR),
+		.size_id = io_read32(base + _RISAF_SIDR)
+	};
+
+	DMSG("RISAF %#"PRIxPTR" version %d.%d, ip0x%x size0x%x",
+	     risaf->pdata.base.pa, version.major, version.minor, version.ip_id,
+	     version.size_id);
+}
+
+/*
+ * @brief  Lock the RISAF IP registers for a given instance.
+ * @param  risaf: RISAF instance.
+ * @retval 0 if OK, negative value else.
+ */
+static __maybe_unused
+int stm32_risaf_lock(struct stm32_risaf_instance *risaf)
+{
+	assert(risaf);
+
+	io_setbits32(risaf_base(risaf) + _RISAF_CR, _RISAF_CR_GLOCK);
+
+	return 0;
+}
+
+/*
+ * @brief  Get the RISAF lock state for a given instance.
+ * @param  risaf: RISAF instance.
+ *         state: lock state, true if locked, false else.
+ * @retval 0 if OK, negative value else.
+ */
+static __maybe_unused
+int stm32_risaf_is_locked(struct stm32_risaf_instance *risaf, bool *state)
+{
+	assert(risaf);
+
+	*state = (io_read32(risaf_base(risaf) + _RISAF_CR) &
+		  _RISAF_CR_GLOCK) == _RISAF_CR_GLOCK;
+
+	return 0;
+}
+
+static TEE_Result stm32_risaf_init_ddata(struct stm32_risaf_instance *risaf)
+{
+	vaddr_t base = risaf_base(risaf);
+	uint32_t granularity = 0;
+	uint32_t mask_lsb = 0;
+	uint32_t mask_msb = 0;
+	uint32_t hwcfgr = 0;
+
+	risaf->ddata = calloc(1, sizeof(*risaf->ddata));
+	if (!risaf->ddata)
+		return TEE_ERROR_OUT_OF_MEMORY;
+
+	/* Get address mask depending on RISAF instance HW configuration */
+	hwcfgr =  io_read32(base + _RISAF_HWCFGR);
+	mask_lsb = (hwcfgr & _RISAF_HWCFGR_CFG3_MASK) >>
+		   _RISAF_HWCFGR_CFG3_SHIFT;
+	mask_msb = mask_lsb + ((hwcfgr & _RISAF_HWCFGR_CFG4_MASK) >>
+			       _RISAF_HWCFGR_CFG4_SHIFT) - 1U;
+	risaf->ddata->mask_regions = GENMASK_32(mask_msb, mask_lsb);
+	risaf->ddata->max_base_regions = (hwcfgr & _RISAF_HWCFGR_CFG1_MASK) >>
+					 _RISAF_HWCFGR_CFG1_SHIFT;
+
+	/* Get IP region granularity */
+	granularity = io_read32(risaf_base(risaf) + _RISAF_HWCFGR);
+	granularity = BIT((granularity & _RISAF_HWCFGR_CFG3_MASK) >>
+			  _RISAF_HWCFGR_CFG3_SHIFT);
+	risaf->ddata->granularity = granularity;
+
+	return TEE_SUCCESS;
+}
+
+TEE_Result stm32_risaf_reconfigure(paddr_t base)
+{
+	struct stm32_risaf_instance *risaf = NULL;
+
+	SLIST_FOREACH(risaf, &risaf_list, link) {
+		struct stm32_risaf_region *regions = risaf->pdata.regions;
+		TEE_Result res = TEE_ERROR_GENERIC;
+		unsigned int i = 0;
+
+		if (base != risaf->pdata.base.pa)
+			continue;
+
+		res = clk_enable(risaf->pdata.clock);
+		if (res)
+			return res;
```

Comment:
Failing here would leave the platform partially configured. I think panicking here would be simpler.